core/drivers/imx_uart.c

```diff
@@ -110,11 +113,14 @@ static void imx_uart_putc(struct serial_chip *chip, int ch)
 {
 	vaddr_t base = chip_to_base(chip);
 
-	write32(ch, base + UTXD);
+	if (!(read32(base + UCR1) & UCR1_UARTEN))
+		return;
 
-	/* Wait until sent */
-	while (!(read32(base + UTS) & UTS_TXEMPTY))
+	/* Wait until there's space in the TX FIFO */
+	while (read32(base + UTS) & UTS_TXFULL)
```

Comment:
It will hang.

Reply:
It seems rather fragile for secure world to depend on normal world for this particular behavior. Do you have any objections to skipping UART write if UART is disabled? It looks like the other serial drivers have this behavior.

Reply:
Currently, for Linux + OP-TEE, when i debug suspend/resume, I normally first enable no_console_suspend for linux and enable debug for OP-TEE; then remove no_console_suspend for linux and remove debug for OP-TEE, if still issues, I use jtag.

Because of the hardware design, the best approach for handling clock/power/rtc and etc stuffs should be implementing them in secure world, but because of ARMv7 limitation, those codes needs be in OP-TEE repo, and requires huge efforts for normal world and OP-TEE.

I do not have better ideas, except implementing the above in OP-TEE.

Reply:
@jenswi-linaro There is still risk, because during the fifo check loop, there is still possiblility the normal world could disable the uart, anyway this depends on normal world uart usage.

Not only UART, for other devices, if linux disables the clock, OP-TEE will also hang when accessing the devices. Just like what most ARMv8 processor implemented using a dedicate processor for clock/power, for ARMv7, secure world could take the responsibility to handle those.

Reply:
@MrVan I see your point and thanks for the clarification.

With the current design OP-TEE cannot protect against suddenly stopped clocking or shut down power, but we can at least check that the UART is enabled while waiting for room in the FIFO.

Reply:
@MrVan @jenswi-linaro I moved the UART disabled check inside the loop, so if UART is disabled partway through write it will not hang. There should be no performance impact because the preceding change should ensure FIFO sits full of characters most of the time.